src/ape/managers/chain.py

```diff
@@ -835,6 +841,33 @@ def get_proxy_info(self, address: AddressType) -> Optional[ProxyInfoAPI]:
 
         return self._local_proxies.get(address) or self._get_proxy_info_from_disk(address)
 
+    def get_creation_metadata(self, address: AddressType) -> Optional[ContractCreation]:
+        """
+        Get contract creation metadata containing txn_hash, deployer, factory, block.
+
+        Args:
+            address (AddressType): The address of the contract.
+
+        Returns:
+            Optional[:class:`~pae.api.query.ContractCreation`]
+        """
+        if creation := self._local_contract_creation.get(address):
+            return creation
+        # read from disk
+        if creation := self._get_contract_creation_from_disk(address):
+            self._local_contract_creation[address] = creation
+            return creation
+        # query and cache
+        query = ContractCreationQuery(columns=["*"], contract=address)
+        try:
+            creation = next(self.query_manager.query(query))
+        except StopIteration:
+            return None
```

Comment:
Think it's just better to raise a normal exception that tells you the algo is broken and to raise an issue in ape repo

Do we have one of those "we don't expect it to fail, but if it did please create an issue" type of exceptions?

Reply:
a `ValueError` works for me, as this is unlikely. Anything is better than `StopIteration`.

Reply:
Vyper has this special exception, it wouldn't subclass `ApeException` but would be a special exception you just put anywhere you think you're safe but just to be sure (or protect future maintainers from stupidity)

Maybe `ape.exceptions.Unreachable`?

Reply:
I really don't like the name `Unreachable`. Everything is reachable pretty much. You may think it is not, but damn it, I will somehow reach it!!

Reply:
Just raise  an exception saying "For some unknown reason, there is no creation block."

Reply:
I mean - I could create a contract after deploying, then revert to some older block, delete stuff from the cache or something, and then query a creation block and get nowhere.

Reply:
Could even be the `ape.exceptions.QueryError`

Reply:
actually, is `None` so bad?

Reply:
> Just raise an exception saying "For some unknown reason, there is no creation block."

Well the algo is basically not ever supposed to return None, if the algo is correct. If the algo is wrong, it may. But if it's wrong, that's a guaranteed bug

Reply:
"the algo" meaning the binary search function

Reply:
How does the algo handle time travel though?

Reply:
not sure what that means. the assumption here is that a contract is only created one time, and that you have the block available where it is created (otherwise it (should) raise)

Reply:
time travel refers to snapshotting and reverting. You can call a method anytime, You can clear a cache anytime.

Reply:
i think returning None here is consistent with other similar methods like `get_proxy_info`. the query would return None for non-contracts.